### PR TITLE
Fix cli print out completeness and cyclic handling issue

### PR DESCRIPTION
**Pull request by wisechengyi**
## Problem
Before resolve/fetch does not output the complete graph. E.g.
```
./coursier resolve -t org.apache.avro:avro:1.7.4 org.apache.commons:commons-compress:1.4.1
 Result:
├─ org.apache.avro:avro:1.7.4
│  ├─ com.thoughtworks.paranamer:paranamer:2.3
│  ├─ org.apache.commons:commons-compress:1.4.1
│  │  └─ org.tukaani:xz:1.0
│  ├─ org.codehaus.jackson:jackson-core-asl:1.8.8
│  ├─ org.codehaus.jackson:jackson-mapper-asl:1.8.8
│  │  └─ org.codehaus.jackson:jackson-core-asl:1.8.8
│  ├─ org.slf4j:slf4j-api:1.6.4
│  └─ org.xerial.snappy:snappy-java:1.0.4.1
└─ org.apache.commons:commons-compress:1.4.1
          // why is there nothing over here?
```

## Solution
Do not avoid jars have been printed out before. E.g.
```
]$ cli/target/pack/bin/coursier resolve  -t org.apache.avro:avro:1.7.4   org.apache.commons:commons-compress:1.4.1 
  Result:
├─ org.apache.avro:avro:1.7.4
│  ├─ com.thoughtworks.paranamer:paranamer:2.3
│  ├─ org.apache.commons:commons-compress:1.4.1
│  │  └─ org.tukaani:xz:1.0
│  ├─ org.codehaus.jackson:jackson-core-asl:1.8.8
│  ├─ org.codehaus.jackson:jackson-mapper-asl:1.8.8
│  │  └─ org.codehaus.jackson:jackson-core-asl:1.8.8
│  ├─ org.slf4j:slf4j-api:1.6.4
│  └─ org.xerial.snappy:snappy-java:1.0.4.1
└─ org.apache.commons:commons-compress:1.4.1
   └─ org.tukaani:xz:1.0
```

### CODE REVIEW

**Comment by alexarchambault** [comment]
Thanks! I'm now realizing that https://github.com/coursier/coursier/pull/376, which fixed https://github.com/coursier/coursier/issues/266, added no non regression test for it... And I'm not sure the changes here are fine with it.

I think the code in question should simply remember what are the parents of the node about to be printed, and not print it if it has itself has a parent (which the current code does, but in a buggy way).

**Comment by wisechengyi** [comment]
haha yeah I broke it again :D 
```
$ coursier resolve -t \
    -r http://cogcomp.cs.illinois.edu/m2repo \
    edu.illinois.cs.cogcomp:illinois-nlp-pipeline:0.1.21
```

**Comment by wisechengyi** [comment]
To clarify what you mean by 
>  should simply remember what are the parents of the node about to be printed, and not print it if it has itself has a parent

Basically it is to see if any direct ancestor has appeared before, if so, do not recurse.

if 
```
a -> b -> c -> a
            -> e -> f
```
```
resolve a, e
```
should give:
```
a 
-b
--c 
// not printing 'a' again
---e
----f
e
-f
```

@alexarchambault could you kindly confirm?

**Comment by alexarchambault** [comment]
@wisechengyi Yep, that's what I meant!

**Comment by wisechengyi** [comment]
Reverting to the style pre https://github.com/coursier/coursier/pull/376/files, and add cycle detection.

**Comment by wisechengyi** [comment]
hey @alexarchambault, do you know if there is a way to trigger a single test from sbt, like `coursier.util.TreeTests`? still scala newbie here

**Comment by alexarchambault** [comment]
@wisechengyi Something like `testsJVM/testOnly coursier.util.TreeTests` at the sbt prompt, should do, like
```
$ sbt
…
> testsJVM/testOnly coursier.util.TreeTests
…
> ~testsJVM/testOnly coursier.util.TreeTests
…
```

(second one watches for changes)

**Comment by alexarchambault** [comment]
@wisechengyi Tell me as soon as you want me to review.

**Comment by wisechengyi** [comment]
@alexarchambault should be good for review now. thank you!

File: core/shared/src/main/scala/coursier/util/Tree.scala [hunk 1]
```diff
@@ -16,8 +16,9 @@ object Tree {
       */
     def recursivePrint(elems: Seq[T], ancestors: Set[T], prefix: String, acc: String => Unit): Unit = {
       val unseenElems: Seq[T] = elems.filterNot(ancestors.contains)
-      for (elem <- unseenElems) {
-        val isLast = unseenElems.indexOf(elem) == unseenElems.length - 1
```
Comment (alexarchambault) [inline core/shared/src/main/scala/coursier/util/Tree.scala]: I think the call to `indexOf` makes things quadratic here (calling `indexOf` for each element should be quadratic wrt to `unseenElems.length`). I pushed a quick fix.

Comment (wisechengyi) [reply core/shared/src/main/scala/coursier/util/Tree.scala]: ok good catch. thanks!

**Comment by alexarchambault** [comment]
Merging, thanks!

**Pull request merged**